app/src/main/java/com/x8bit/bitwarden/data/autofill/accessibility/manager/AccessibilityFieldManagerImpl.kt

```diff
@@ -0,0 +1,84 @@
+
+private const val MAX_NODE_COUNT: Int = 100
+private const val MAX_RECURSION_DEPTH: Int = 100
+
+/**
+ * The default implementation for the [AccessibilityFieldManager].
+ */
+class AccessibilityFieldManagerImpl : AccessibilityFieldManager {
+    override fun findAccessibilityNodeInfos(
+        rootNode: AccessibilityNodeInfo,
+        recursionDepth: Int,
+        predicate: (AccessibilityNodeInfo) -> Boolean,
+    ): List<AccessibilityNodeInfo> {
+        if (predicate(rootNode)) return listOf(rootNode)
+        if (recursionDepth >= MAX_RECURSION_DEPTH) return emptyList()
+        val childNodeCount = rootNode.childCount - 1
+        if (childNodeCount > MAX_NODE_COUNT) log(message = "Too many child iterations.")
+        return (0..childNodeCount.coerceAtMost(maximumValue = MAX_NODE_COUNT)).flatMap {
+            val childNode = rootNode.getChild(it) ?: return@flatMap emptyList()
+            if (childNode.hashCode() == this.hashCode()) {
+                log(message = "Child node is the same as parent for some reason.")
+                emptyList()
+            } else {
+                findAccessibilityNodeInfos(
+                    rootNode = childNode,
+                    recursionDepth = recursionDepth + 1,
+                    predicate = predicate,
+                )
+            }
+        }
+    }
+
+    override fun findUsernameField(
+        uri: Uri,
+        allFields: List<AccessibilityNodeInfo>,
+        passwordFields: List<AccessibilityNodeInfo>,
+    ): AccessibilityNodeInfo? {
+        val uriPath = uri
+            .path
+            ?: return findMissingUsernameField(
+                allFields = allFields,
+                passwordFields = passwordFields,
+            )
+        return uri
+            .authority
+            ?.removePrefix(prefix = "www.")
+            ?.getKnownUsernameFieldNull()
+            ?.let { usernameField ->
+                allFields.firstOrNull { node ->
+                    node.isUsername(
+                        uriPath = uriPath,
+                        knownUsernameField = usernameField,
+                    )
+                }
+            }
+            ?: findMissingUsernameField(allFields = allFields, passwordFields = passwordFields)
+    }
+
+    /**
+     * Attempts to find a username field if there isn't one already. This functions by finding the
+     * first known password field and taking the field directly above it.
+     */
+    private fun findMissingUsernameField(
+        allFields: List<AccessibilityNodeInfo>,
+        passwordFields: List<AccessibilityNodeInfo>,
+    ): AccessibilityNodeInfo? =
+        // Return the field directly above the first password field if possible
+        passwordFields
+            .firstOrNull()
+            ?.let { allFields.getOrNull(index = allFields.indexOf(element = it) - 1) }
+
+    private fun log(message: String) {
+        if (!BuildConfig.DEBUG) return
+        Log.i("AccessibilityFieldManager", message)
```

Comment:
💭 def outside scope of this, but maybe we consider adding something like timber in the future to make this testable.

Reply:
That has a whole approval process to go through but there is a ticket for that.